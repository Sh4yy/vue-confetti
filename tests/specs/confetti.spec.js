import Confetti from '../../src/confetti';

expect.extend({
  toHaveCanvasEvents(canvas) {
    // eslint-disable-next-line no-underscore-dangle
    const hasEvents = canvas.getContext('2d').__getEvents().length > 0;

    return {
      pass: hasEvents,
      message: () => `Expected canvas to${this.isNot ? ' not ' : ' '}have events`,
    };
  },
});

describe('Confetti', () => {
  beforeEach(() => {
    // Mock for a single frame only
    jest.spyOn(window, 'requestAnimationFrame').mockImplementationOnce(cb => cb());
  });

  afterEach(() => {
    window.requestAnimationFrame.mockRestore();
  });

  it('is initialised with default settings', () => {
    const confetti = new Confetti();

    expect(confetti).toMatchSnapshot();
  });

  it('does not throw when started without options', () => {
    const confetti = new Confetti();

    expect(() => confetti.start()).not.toThrow();
  });

  describe('start', () => {
    it('accepts a canvas by element reference', () => {
      const confetti = new Confetti();
      const canvasElement = document.createElement('canvas');

      confetti.start({ canvasElement });

      expect(canvasElement).toHaveCanvasEvents();
    });

    it('accepts a canvas by ID', () => {
      const confetti = new Confetti();
      const canvasElement = document.createElement('canvas');

      canvasElement.id = 'my-canvas';
      document.body.appendChild(canvasElement);

      confetti.start({ canvasId: 'my-canvas' });

      expect(canvasElement).toHaveCanvasEvents();
    });

    it('throws if given element is not a canvas', () => {
      const confetti = new Confetti();
      const canvasElement = document.createElement('div');

      expect(() => confetti.start({ canvasElement })).toThrow(
        /.*not a valid HTMLCanvasElement.*/,
      );
    });

    it('throws if given ID is not for a canvas', () => {
      const confetti = new Confetti();
      const canvasElement = document.createElement('div');
      canvasElement.id = 'not-a-canvas';

      expect(() => confetti.start({ canvasId: 'not-a-canvas' })).toThrow(
        /.*not a valid HTMLCanvasElement.*/,
      );
    });

    it('throws if both canvas element and id are given', () => {
      const confetti = new Confetti();
      const canvasElement = document.createElement('canvas');

      expect(() => confetti.start({ canvasId: 'foo', canvasElement })).toThrow(
        /.*mutually exclusive.*/,
      );
    });
  });

  describe('update', () => {
    it('restarts if given a new canvas by element reference', () => {
      const confetti = new Confetti();
      const canvasElementOne = document.createElement('canvas');
      const canvasElementTwo = document.createElement('canvas');

      const spy = jest.spyOn(confetti, 'start');

      confetti.start({ canvasElement: canvasElementOne });
      confetti.update({ canvasElement: canvasElementTwo });

      expect(spy).toHaveBeenCalledWith({ canvasElement: canvasElementTwo });
    });

    it('restarts if given a new canvas by ID', () => {
      const confetti = new Confetti();
      const canvasElementOne = document.createElement('canvas');
      const canvasElementTwo = document.createElement('canvas');

      const spy = jest.spyOn(confetti, 'start');

      canvasElementOne.id = 'my-canvas-one';
      canvasElementTwo.id = 'my-canvas-two';
      document.body.appendChild(canvasElementOne);
      document.body.appendChild(canvasElementTwo);

      confetti.start({ canvasId: 'my-canvas-one' });
      confetti.update({ canvasId: 'my-canvas-two' });

      expect(spy).toHaveBeenCalledWith({ canvasId: 'my-canvas-two' });
    });
  });
});
