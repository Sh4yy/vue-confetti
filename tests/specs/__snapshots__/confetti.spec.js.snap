// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confetti is initialised with default settings 1`] = `
Confetti {
  "H": 0,
  "W": 0,
  "animationId": null,
  "canvas": null,
  "canvasEl": null,
  "framesSinceDrop": 0,
  "killed": false,
  "particleManager": null,
  "particlesPerFrame": 0,
  "wind": 0,
  "windChange": 0.01,
  "windPosCoef": 0.002,
  "windSpeed": 1,
  "windSpeedMax": 1,
}
`;
